ESQL: INLINESTATS implementation with multiple LogicalPlan updates

astefan opened the pull request:
Part of https://github.com/elastic/elasticsearch/issues/124715 and similar to https://github.com/elastic/elasticsearch/pull/128476.
Different from https://github.com/elastic/elasticsearch/pull/128476 in that it takes a "LogicalPlan" approach to running a sub-query, integrating its result back in the "main" LogicalPlan and continuing running the query.

astefan committed:
Adopt a "LogicalPlan" approach to running multiple sub-queries,
coordinated from the EsqlSession.

astefan committed:
Update docs/changelog/128917.yaml

astefan committed:
checkstyle

astefan committed:
More tests un-Ignored and updates after update from main

astefan committed:
Fix capability

astefan committed:
TEMP

astefan committed:
Adapt PruneColumns to have "visibility" into the right-hand side of
the inlinestats JOIN

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into support_multi_inlinestats_logicalPlan_approach

astefan committed:
More fixes and tests

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into support_multi_inlinestats_logicalPlan_approach

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/InlineJoin.java:
```diff
@@ -139,20 +147,12 @@ public List<Attribute> computeOutput(List<Attribute> left, List<Attribute> right
-            // as first columns in the output followed by whatever the right hand side of join adds in this order: aggregates first,
-            // followed by groupings (this order should be preserved inside the rightFields() output)
-            output = mergeOutputAttributes(right, leftOutputWithoutMatchFields);
+            List<Attribute> leftOutputWithoutKeys = left.stream().filter(attr -> config().leftFields().contains(attr) == false).toList();
```
This is taken from @alex-spies great suggestion [here](https://github.com/elastic/elasticsearch/pull/123589/files/b2c34742317c64c21a9b5019bd642cdb69178353#r1979692509)

astefan committed:
More deterministic test

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into support_multi_inlinestats_logicalPlan_approach

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

astefan committed:
Update docs/changelog/128917.yaml

elasticsearchmachine commented:
Hi @astefan, I've updated the changelog YAML for you.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PruneColumns.java:
```diff
@@ -131,14 +139,15 @@ public LogicalPlan apply(LogicalPlan plan) {
             N prev = it.previous();
-            if (used.contains(prev.toAttribute()) == false) {
+            var attr = prev.toAttribute();
+            if (used.contains(attr) == false && exceptions.contains(attr) == false) {
```
This was preventing queries like https://github.com/elastic/elasticsearch/pull/128917/files#diff-a5ed793b796c14ec1f92807242ff2376e16888c8cf7627a4f6a8f57f23ed843aR1061-R1115.

astefan committed:
Test inlinestats after stats with bucket and top

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into support_multi_inlinestats_logicalPlan_approach

astefan committed:
Merge branch 'support_multi_inlinestats_logicalPlan_approach' of https://github.com/astefan/elasticsearch into support_multi_inlinestats_logicalPlan_approach

astefan committed:
Have Row - LocalRelation transformation use a copying blocks approach

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into support_multi_inlinestats_logicalPlan_approach

astefan committed:
Attempt a serialization change to have to Suppliers for LocalRelation

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into support_multi_inlinestats_logicalPlan_approach

alex-spies reviewed (commented):
I had a first look, focusing on the added tests. Very nice!

astefan committed:
Adapt to EXPLAIN command

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into support_multi_inlinestats_logicalPlan_approach

bpintea reviewed (commented):
A fist pass, minor remarks, will conclude tomorrow.

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -231,85 +231,88 @@ public void executeOptimizedPlan(
     ) {
-        PlanTuple tuple = subPlanIterator.next();
+        // Create a physical plan out of the logical sub-plan
+        var physicalSubPlan = logicalPlanToPhysicalPlan(subPlans.nonStubbedSubPlan, request);
```
Abs. optional, but these logs helped a bit go through the execution.
```
diff --git a/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java b/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java
index d8118b0e0f4e..7fac395ddfb4 100644
--- a/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java
+++ b/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java
@@ -285,6 +285,7 @@ public class EsqlSession {
         EsqlQueryRequest request,
         ActionListener<Result> listener
     ) {
+        LOGGER.debug("Executing subplan:\n{}", subPlans.nonStubbedSubPlan);
         // Create a physical plan out of the logical sub-plan
         var physicalSubPlan = logicalPlanToPhysicalPlan(subPlans.nonStubbedSubPlan, request);

@@ -300,10 +301,12 @@ public class EsqlSession {
                     ij -> ij.right() == subPlans.originalSubPlan ? InlineJoin.inlineData(ij, resultWrapper) : ij
                 );
                 newLogicalPlan.setOptimized();
+                LOGGER.debug("Plan after previous subplan execution:\n{}", newLogicalPlan);
                 // look for the next inlinejoin plan
                 var newSubPlan = firstSubPlan(newLogicalPlan);

                 if (newSubPlan == null) {// run the final "main" plan
+                    LOGGER.debug("Executing final plan:\n{}", newLogicalPlan);
                     var newPhysicalPlan = logicalPlanToPhysicalPlan(newLogicalPlan, request);
                     runner.run(newPhysicalPlan, next.delegateFailureAndWrap((finalListener, finalResult) -> {
                         completionInfoAccumulator.accumulate(finalResult.completionInfo());
```

bpintea reviewed (approved):
Nice work, LGTM.

My only significant note, not directly PR-specific, is having INLINESTATS hit the same source (i.e. with same filters, sorting, etc. that come before the first INLINESTATS) multiple times (i.e. for each INLINESTATS). It'd be great if we could save this computation and reuse it, that might become heavy otherwise if there's denser logic than just Lucene (like more LOOKUP JOINs, ENRICHes, and then more INLINESTATs). I guess we could look into this at some later point.

alex-spies reviewed (commented):
Second round - focused on the execution in EsqlSession this time.

Looks like we're going back in the direction of the initial approach to inline stats, where each phase had a full optimizer run. That's the right approach IMO.

I want to have another look next week to better understand the changes to the optimizer rules. But please go ahead and merge whenever you're happy (I see Bogdan already 👍'd the PR); if I find something I find important, I'll leave a comment after the merge.

astefan commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/inlinestats.csv-spec:
```diff
@@ -1001,3 +1108,171 @@ FROM hosts METADATA _index
 alpha db server    |alpha          |127.0.0.1      |hosts           |127.0.0.1|1         |DB servers       |eth0
 ;
+
+doubleShadowing
```
The "no groups" part is the problematic one **now** (as in what I am still working on as part of the same PR that's still evolving).

bpintea replied:
I did give it a try and it worked fine. But haven't gone crazy with it, so more work is welcome, if needed.

astefan replied:
I added the fix in the `PruneColumns` rule. It's about removing the `inlinestats` completely if its output is not needed. What happened in some queries was that the aggregation of the `inlinestats` was removed and the `inlinejoin` was left in a completely messed up state.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -231,85 +231,88 @@ public void executeOptimizedPlan(
             values.add(List.of("coordinator", "optimizedPhysicalPlan", physicalPlanString));
```
Yes, this PR wants to be a solid base for further parallel branches of work on `inlinestats`. Modelling the InlineJoin properly is one of those follow up things.

astefan commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/inlinestats.csv-spec:
```diff
@@ -212,14 +212,41 @@ FROM employees
+
+three
+required_capability: inlinestats_v8
+// fails with AssertionError at org.elasticsearch.xpack.esql.plan.logical.Limit.writeTo(Limit.java:70)
```
This used to happen. Haven't seen it anymore. I kept the comment (slightly inaccurate) so that I remember this aspect. I knew about `limit`, but at the time this was happening I didn't dig into it enough to understand if it's still ok to happen for inlinestats, since there are "things" that are reused.

I've reworded (used past tense) the comment and kept it there.

astefan commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/inlinestats.csv-spec:
The optimized logicalplan looks like this for that query you suggested (`from employees | sort emp_no desc | inlinestats avg = avg(salary) by languages | limit 5 | keep emp_no, avg, languages, gender`):

```
EsqlProject[[emp_no{f}#411, avg{r}#400, languages{f}#407, gender{f}#406]]
 \_Limit[5[INTEGER],true]
   \_InlineJoin[LEFT,[languages{f}#407],[languages{f}#407],[languages{r}#407]]
     |_TopN[[Order[emp_no{f}#411,DESC,FIRST]],5[INTEGER]]
     | \_EsRelation[employees][emp_no{f}#411, gender{f}#406, languages{f}#407, lan..]
     \_Project[[avg{r}#400, languages{f}#407]]
       \_Eval[[$$SUM$avg$0{r$}#413 / $$COUNT$avg$1{r$}#414 AS avg#400]]
         \_Aggregate[[languages{f}#407],[SUM(salary{f}#412,true[BOOLEAN]) AS $$SUM$avg$0#413, COUNT(salary{f}#412,true[BOOLEAN]) AS $$COUNT$avg$1#414, languages{f}#407]]
           \_StubRelation[[emp_no{f}#411, gender{f}#406, languages{f}#407, languages.byte{f}#408, languages.long{f}#410, languages.short{f}#409, salary{f}#412]]
```

which I think it looks right:
- the right hand side is untouched
- the limit is pushed down on the left hand side and made up as TopN
- final limit on top of the inlinestats

I will add a TODO in the csv-spec for this exact query so that we also have a unit test for this.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/local/CopyingLocalSupplier.java:
```diff
@@ -0,0 +1,31 @@
+/**
+ * A {@link LocalSupplier} that allways creates a new copy of the {@link Block}s initially provided at creation time.
+ */
+public class CopyingLocalSupplier extends ImmediateLocalSupplier {
```
Yeah, it's about that and is really specific to `row` queries. I'll add a comment. I am not entirely happy with how this came up and it's likely to change in the future, but the comment will live on.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/SessionUtils.java:
```diff
@@ -26,7 +26,9 @@ public static Block[] fromPages(List<Attribute> schema, List<Page> pages) {
         if (bytesUsed > ByteSizeValue.ofMb(1).getBytes()) {
-            throw new IllegalArgumentException("first phase result too large [" + ByteSizeValue.ofBytes(bytesUsed) + "] > 1mb");
+            throw new IllegalArgumentException(
+                "INLINESTATS sub-plan execution results too large [" + ByteSizeValue.ofBytes(bytesUsed) + "] > 1mb"
```
Good point. I'll change it.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -207,85 +206,85 @@ public void executeOptimizedPlan(
     }
 
-    private record PlanTuple(PhysicalPlan physical, LogicalPlan logical) {}
+    private record LogicalPlanTuple(LogicalPlan nonStubbedSubPlan, LogicalPlan originalSubPlan) {}
```
Yeah, the structure of these classes is kind of chaotic now. I regard this still as preliminary work. At some point this should be, hopefully, abstracted enough to be reused for subqueries as well.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -217,8 +217,8 @@ public void executeOptimizedPlan(
         ActionListener<Result> listener
     ) {
-        PhysicalPlan physicalPlan = logicalPlanToPhysicalPlan(optimizedPlan, request);
         if (explainMode) {
```
I've added a TODO to revisit this. `explain` in EsqlSession needs a second look, some things do not feel like they belong here.

astefan committed:
Add more tests and fix incorrect pruning of columns of the inlinestats'
Aggregation.
Move two methods from EsqlSession to InlineJoin.
Address reviews
Add more comments

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into support_multi_inlinestats_logicalPlan_approach

astefan committed:
More stuff

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into support_multi_inlinestats_logicalPlan_approach

alex-spies reviewed (approved):
Heya, all done now.

This is a nice advancement of INLINE STATS, let's go! I think there are a couple things to be looked at in follow ups, but this is fine to merge as-is IMO; we can continue iterating in follow-ups as the changes here do not seem to affect the code health of the non-INLINE STATS code paths.

Most notable follow-up items IMO:
- Correctness of optimization rules w.r.t. stubs, esp. the limit pushdown.
- More logical plan optimizer tests for PruneColumns and a good hard look at PruneColumns (maybe I'm misreading things though, see comments below)

astefan committed:
Tests fixes

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into support_multi_inlinestats_logicalPlan_approach

astefan commented:
Asking some feedback from @nik9000 as well, especially related to CopyingLocalSupplier and whatever I am doing in there related to Blocks :-). I didn't change the ImmediateLocalSupplier to always behave like CopyingLocalSupplier because I wanted this behavior only for `row` commands where we know that blocks will always be released on a single node (the coordinator) whereas a LocalRelation **might** exist also on data nodes. I wanted to separate these two use cases completely.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PruneColumns.java:
```diff
@@ -63,14 +67,18 @@ public LogicalPlan apply(LogicalPlan plan) {
             }
 
+            if (p instanceof InlineJoin ij) {
+                inlineJoinRightOutput.addAll(ij.right().outputSet());
```
Yes, this will be followups.

astefan commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/inlinestats.csv-spec:
```diff
@@ -245,6 +251,25 @@ emp_no:integer |avg_worked_seconds:long|avg_avg_worked_seconds:double|languages:
+;
+
+ emp_no:integer| avg:double    |languages:integer|gender:keyword     
+10001          |57305.0        |2                |M              
```
I think this is a bit overkill, tbh. Whoever wants to really dig into the details **will test** this in a real use case.

astefan committed:
Further fixes and addressing reviews

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into support_multi_inlinestats_logicalPlan_approach

nik9000 reviewed (changes requested):
I have some opinions on the serialization change. Compared to the rest of the work they are super minor, but I think we should resolve my questions before merging.

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into support_multi_inlinestats_logicalPlan_approach

astefan committed:
Rework the LocalSupplier serialization

astefan committed:
Address reviews

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into support_multi_inlinestats_logicalPlan_approach

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into support_multi_inlinestats_logicalPlan_approach

nicpenning commented:
Would Inline stats provide the result that this issue presents?

https://github.com/elastic/elasticsearch/issues/128612
